src/cca/scripts/uproxy_server.ts

```diff
@@ -56,25 +36,146 @@ export class UproxyServer implements Server {
   }
 }
 
+// https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
+function getIsStorageAvailable() {
+  try {
+    const storage = window['localStorage'];
+    const x = '__storage_test__';
+    storage.setItem(x, x);
+    storage.removeItem(x);
+    return true;
+  } catch(e) {
+    return false;
+  }
+}
+const isStorageAvailable = getIsStorageAvailable();
+
+// Name by which servers are saved to storage.
+const CONTACTS_STORAGE_KEY = 'cloud-server-invites';
+
 export class UproxyServerRepository implements ServerRepository {
+  // core must already be logged into social networks.
   constructor(private core: CoreConnector, private vpnDevice: VpnDevice) {
-    this.core.login({
-      network: 'Cloud',
-      loginType: uproxy_core_api.LoginType.INITIAL,
+    if (!isStorageAvailable) {
+      console.warn('localStorage unavailable - contacts will not be saved');
+      return;
+    }
+  }
+
+  public onServerCallback :OnServerCallback;
+  public onServer(callback:OnServerCallback) {
+    this.onServerCallback = callback;
+    return this;
+  }
+
+  public restore() {
+    if (!this.onServerCallback) {
+      throw new Error('must call onServer first');
+    }
+
+    for (const savedContact of this.loadContacts().contacts) {
+      this.informCoreOfServer(savedContact.invite).then((server) => {
+        try {
+          this.onServerCallback(server);
+        } catch (e) {
+          console.warn('onServer callback threw', e);
+        }
+      }, (e) => {
+        console.warn('could not inform core of saved server', e);
+      });
+    }
+  }
+
+  // returns saved contacts, which will be empty if
+  // localStorage is unavailable or no contacts are found.
+  // throws if there is an error reading from localStorage or
+  // the stored contacts could not be parsed.
+  private loadContacts(): cloud_social_provider.SavedContacts {
+    if (!isStorageAvailable) {
+      return {
+        contacts: []
+      };
+    }
+
+    const savedContactsJson = localStorage.getItem(CONTACTS_STORAGE_KEY);
+    return savedContactsJson ? <cloud_social_provider.SavedContacts>JSON.parse(
+        savedContactsJson) : {
+      contacts: []
+    };
+  }
+
+  // merges a contact with those already saved to storage.
+  // returns true if the contact was not previously in storage.
+  // contacts are saved as a SavedContacts serialised as JSON. 
+  private saveContact(newInvite:cloud_social_provider.Invite) {
+    if (!isStorageAvailable) {
+      return true;
+    }
+
+    const savedContacts = this.loadContacts();
+
+    // merge or append, as necessary.
+    const dupeIndex = _.findIndex(savedContacts.contacts, (savedContact) => {
```

Comment:
An alternative to this is to have contacts be a host -> info map, so you can do:

    savedContacts.contacts[newSavedContact.invite.host] = newSavedContact

I'd actually decouple the notion of an id from `host`, and have:

    savedContacts.contacts[newSavedContact.id] = newSavedContact

Where the id just happens to be the host address in the case of cloud servers. This decoupling clarifies the concepts and allows us to more easily change in the future.